test/Rentals.spec.ts

```diff
@@ -1569,21 +1569,81 @@ describe('Rentals', () => {
+      ]
+
       await rentals.connect(deployer).initialize(owner.address, mana.address, collector.address, fee)
     })
 
-    it('should allow the asset transfer from a rent', async () => {
+    it('should allow the asset transfer from accepting an offer', async () => {
       expect(await land.ownerOf(tokenId)).to.equal(lessor.address)
 
       await rentals.connect(lessor).acceptOffer({ ...offerParams, signature: await getOfferSignature(tenant, rentals, offerParams) })
 
       expect(await land.ownerOf(tokenId)).to.equal(rentals.address)
     })
 
-    it('should revert when the contract receives an asset not transfered via rent', async () => {
-      const transfer = land.connect(lessor)['safeTransferFrom(address,address,uint256)'](lessor.address, rentals.address, tokenId)
-      await expect(transfer).to.be.revertedWith('Rentals#onERC721Received: ONLY_ACCEPT_TRANSFERS_FROM_THIS_CONTRACT')
+    it('should allow the asset transfer from accepting a listing', async () => {
+      expect(await land.ownerOf(tokenId)).to.equal(lessor.address)
+
+      await rentals
+        .connect(tenant)
+        .acceptListing(
+          { ...listingParams, signature: await getListingSignature(lessor, rentals, listingParams) },
+          acceptListingParams.operator,
+          acceptListingParams.index,
+          acceptListingParams.rentalDays,
+          acceptListingParams.fingerprint
+        )
+
+      expect(await land.ownerOf(tokenId)).to.equal(rentals.address)
+    })
+
+    it('should accept an offer by transfering the asset to the rentals contract with the offer data', async () => {
+      expect(await land.ownerOf(tokenId)).to.equal(lessor.address)
+
+      let rental = await rentals.rentals(offerParams.contractAddress, offerParams.tokenId)
+
+      expect(rental.lessor).to.equal(zeroAddress)
+      expect(rental.tenant).to.equal(zeroAddress)
+      expect(rental.endDate).to.equal(0)
+
+      const bytes = ethers.utils.defaultAbiCoder.encode([offerEncodeType], [offerEncodeValue])
+
+      await land.connect(lessor)['safeTransferFrom(address,address,uint256,bytes)'](lessor.address, rentals.address, tokenId, bytes)
+
+      expect(await land.ownerOf(tokenId)).to.equal(rentals.address)
+
+      rental = await rentals.rentals(offerParams.contractAddress, offerParams.tokenId)
+
+      const latestBlockTime = await getLatestBlockTimestamp()
+
+      expect(rental.lessor).to.equal(lessor.address)
+      expect(rental.tenant).to.equal(tenant.address)
+      expect(rental.endDate).to.equal(latestBlockTime + daysToSeconds(offerParams.rentalDays))
+    })
+
+    it('should revert when the caller is different from the contract address provided in the offer', async () => {
+      const bytes = ethers.utils.defaultAbiCoder.encode([offerEncodeType], [offerEncodeValue])
+      await expect(rentals.onERC721Received(extra.address, lessor.address, tokenId, bytes)).to.be.revertedWith(
+        'Rentals#onERC721Received: SENDER_CONTRACT_ADDRESS_MISMATCH'
```

Comment:
I would try to add more edge cases and therefore more tests. Also, add the ones for the acceptOffer.

- Try to test every revert reason inside `_acceptOffer`
- Try to revert if the user is safe-transferring the asset for a re-rent while the asset is inside the rent contract
- Try to revert if the token Id being sent is not the one expected.